storage/innobase/trx/trx0rec.cc

```diff
@@ -2193,11 +2201,21 @@ trx_undo_prev_version_build(
 
 	ut_ad(!index->table->skip_alter_undo);
 
-	trx_undo_rec_t*	undo_rec = v_status == TRX_UNDO_CHECK_PURGEABILITY
-		? trx_undo_get_rec_if_purgeable(rec_trx_id, index->table->name,
-						roll_ptr, heap)
-		: trx_undo_get_undo_rec(rec_trx_id, index->table->name,
-					roll_ptr, heap);
+	const trx_undo_rec_t* undo_rec;
+
+	switch (v_status) {
+	case TRX_UNDO_CHECK_PURGE_PAGES:
+	case TRX_UNDO_CHECK_PURGEABILITY:
+		undo_rec = trx_undo_get_rec_if_purgeable(
+			rec_trx_id, index->table->name, roll_ptr, heap,
+			v_status == TRX_UNDO_CHECK_PURGE_PAGES);
```

Comment:
I think that the parameter `heap` must be removed here. We would release the S-latch on the buffer pool page that contains the undo record, and retain a buffer-fix until we are done with `undo_rec`. Similar to aa719b5010c929132b4460b78113fbd07497d9c8 of #2792, due to the append-only nature of undo log pages, old records are safe to access as long as we are holding the `purge_sys.end_latch` or `purge_sys.latch` to protect the view that we are relying on (and therefore preventing the page from being freed at the end of a purge batch). It will not matter if other transactions would write elsewhere in the undo log page while we are only holding a buffer-fix. Once we have copied some contents from the `undo_rec` to the `old_vers`, we can let purge resume.

Note that in the `TRX_UNDO_CHECK_PURGE_PAGES` case we are one of the purge threads, and this part of `purge_sys` will be constant until all purge threads are done. We can access `purge_sys.pages` or look up an undo page without holding any `purge_sys` latch.

Reply:
This was addressed in 0480c59ea943efcbb5aba5b02afa8f7138a0f352.